### PR TITLE
[ROCm][Perf][M3] Triton fp32 router GEMM for decode-sized M

### PR DESCRIPTION
**Pull request by benenzhu**
## Purpose

Follow-up to #54845, which landed the gfx950 low-M fp32 router GEMM. This keeps that kernel's math, API and shape allowlist, and changes two things.

**1. Expert-major programs with an XCD remap.** Workgroups are dispatched round-robin over the 8 XCDs, so with the current `pid_m = pid // N` decode consecutive programs on one die walk different experts and every die ends up streaming the whole `N x K` fp32 weight. The weight is the only sizeable operand here, so that is where the time goes. Decoding expert-major and remapping launch ids into per-XCD contiguous blocks leaves each die reading `N / 8` weight rows.

**2. Retuned launch configs, and a token cap of 128 (64 for the largest weight).** Past 32 tokens the op fell back to `F.linear`, but the kernel keeps winning well beyond that, and a decode step with speculative tokens is `num_seqs * (1 + num_spec)` rows -- 16 sequences with 3 speculative tokens is already M = 64. The config table is re-measured over all five supported shapes for M in 1..256: the row tile follows the token count, capped so the grid still fills 256 CUs (which is what keeps the 8-expert shape at `BLOCK_M=1`) and widened once the weight outgrows one XCD's L2 (among the supported shapes only `(6144, 256)` is that large). That same shape is also the one that gives out first against `F.linear`, so it caps at 64 tokens while the rest go to 128.

The decision stays capture-static (shape, dtype, expert count), so it is HIP-graph safe.

## Test Plan

Correctness: `rocm_fp32_router_gemm` against `F.linear(x.float(), w)` for both activation dtypes, all five supported shapes, every M in 1..32 plus samples up to each shape's cap. `tests/kernels/test_rocm_fp32_router_gemm.py` covers this and grows the token range to match; `tests/kernels/test_gate_linear_rocm_dispatch.py` is unchanged.

Performance: one self-contained script (torch + triton only, both kernels inlined, gist to follow) that builds up to 170 rotated `(x, w)` pairs covering ~1 GB so the fp32 weight is never LLC-resident, captures every variant into its own HIP graph, then replays the variants in turn inside each timing round so clock drift lands on all of them alike. 5 rounds x 300 replays per graph, median reported; the min-max range over the five rounds is within 0.1us of the median for every cell below. Eager per-launch timing bottoms out at ~12us of Python launch overhead and cannot rank these kernels. MI355X, one GPU per shape.

## Test Result

MiniMax-M3, `(hidden_size, num_experts) = (6144, 128)`, us per call. "main today" is what `GateLinear` runs at that token count on current main: the #54845 kernel up to 32 tokens, `F.linear` above it.

| M | main today (us) | this PR (us) | latency reduction |
|---|---|---|---|
| 1 | 2.89 | **2.67** | -0.22us (7.6%) |
| 4 | 3.36 | **3.36** | 0.00us (0.0%) |
| 8 | 5.31 | **3.77** | -1.54us (29.0%) |
| 16 | 5.51 | **4.36** | -1.15us (20.9%) |
| 32 | 6.36 | **5.90** | -0.46us (7.2%) |
| 48 | 14.44 (F.linear) | **7.69** | -6.75us (46.7%) |
| 64 | 14.69 (F.linear) | **8.09** | -6.60us (44.9%) |
| 96 | 15.76 (F.linear) | **10.57** | -5.19us (32.9%) |
| 128 | 15.77 (F.linear) | **12.96** | -2.81us (17.8%) |

Two effects add up past 32 tokens: the retuned kernel itself, and the cap no longer sending those counts to `F.linear`. With #54845's own launch config the kernel would take about 9.9us at M=64 and 16.4us at M=128 on this shape, so the retune and the cap each carry roughly half of the gain there.

bench_script: https://gist.github.com/benenzhu/2e1ccffe32fb42ff4b795fe785925b53
<details>
<summary>More shapes</summary>

Same measurement, the other four supported shapes.

`(3072, 256)`

| M | main today (us) | this PR (us) | latency reduction |
|---|---|---|---|
| 1 | 2.64 | **2.36** | -0.28us (10.6%) |
| 8 | 3.98 | **3.70** | -0.28us (7.0%) |
| 16 | 4.70 | **3.97** | -0.73us (15.5%) |
| 32 | 6.07 | **5.50** | -0.57us (9.4%) |
| 48 | 13.38 (F.linear) | **7.05** | -6.33us (47.3%) |
| 64 | 13.18 (F.linear) | **8.91** | -4.27us (32.4%) |
| 128 | 13.84 (F.linear) | **13.33** | -0.51us (3.7%) |

`(4096, 8)`

| M | main today (us) | this PR (us) | latency reduction |
|---|---|---|---|
| 1 | 1.98 | **1.95** | -0.03us (1.5%) |
| 8 | 2.11 | **2.00** | -0.11us (5.2%) |
| 16 | 2.21 | **2.04** | -0.17us (7.7%) |
| 32 | 2.34 | **2.06** | -0.28us (12.0%) |
| 48 | 9.70 (F.linear) | **2.22** | -7.48us (77.1%) |
| 64 | 9.70 (F.linear) | **2.28** | -7.42us (76.5%) |
| 128 | 10.20 (F.linear) | **3.04** | -7.16us (70.2%) |

`(4096, 192)`

| M | main today (us) | this PR (us) | latency reduction |
|---|---|---|---|
| 1 | 2.72 | **2.42** | -0.30us (11.0%) |
| 8 | 4.49 | **3.43** | -1.06us (23.6%) |
| 16 | 4.87 | **4.49** | -0.38us (7.8%) |
| 32 | 6.57 | **5.83** | -0.74us (11.3%) |
| 48 | 14.61 (F.linear) | **7.67** | -6.94us (47.5%) |
| 64 | 14.67 (F.linear) | **8.26** | -6.41us (43.7%) |
| 128 | 15.64 (F.linear) | **13.55** | -2.09us (13.4%) |

`(6144, 256)` -- the only supported shape whose fp32 weight (6.3 MiB) outgrows one XCD's L2. It loses to `F.linear` from 80 tokens up (20.1 vs 16.1us at M=80), so its cap stays at 64 rather than 128.

| M | main today (us) | this PR (us) | latency reduction |
|---|---|---|---|
| 1 | 3.11 | **2.85** | -0.26us (8.4%) |
| 8 | 5.58 | **4.74** | -0.84us (15.1%) |
| 16 | 6.28 | **6.11** | -0.17us (2.7%) |
| 32 | 9.83 | **9.17** | -0.66us (6.7%) |
| 48 | 15.27 (F.linear) | **12.16** | -3.11us (20.4%) |
| 64 | 14.97 (F.linear) | **13.31** | -1.66us (11.1%) |

No shape regresses at any token count the kernel now accepts.

</details>

Correctness: 172 cases (five shapes x M in 1..128 x bf16/fp32, counts above a shape's cap checked for rejection instead) all match `F.linear(x.float(), w)` within 5e-4 absolute, the tolerance the existing tests use.

Server command (MI355X x4, TP4):

```
export VLLM_ENGINE_READY_TIMEOUT_S=3600
export VLLM_EXECUTE_MODEL_TIMEOUT_SECONDS=1800
export VLLM_USE_BREAKABLE_CUDAGRAPH=0
export VLLM_ROCM_USE_AITER=1
export VLLM_ROCM_USE_AITER_MOE=1
export VLLM_ROCM_USE_AITER_FUSION_SHARED_EXPERTS=1
export VLLM_ROCM_SHUFFLE_KV_CACHE_LAYOUT=1
export VLLM_ROCM_QUICK_REDUCE_QUANTIZATION=INT4
export VLLM_ROCM_QUICK_REDUCE_CAST_BF16_TO_FP16=1
export VLLM_ROCM_QUICK_REDUCE_QUANTIZATION_MIN_SIZE_KB=256

vllm serve amd/MiniMax-M3-MXFP4 \
--served-model-name amd/MiniMax-M3-MXFP4 \
--host 0.0.0.0 --port 8888 \
--tensor-parallel-size 4 \
--trust-remote-code \
--block-size 128 \
--gpu-memory-utilization 0.85 \
--enable-chunked-prefill \
--max-num-batched-tokens 32768 \
--language-model-only \
--enable-prefix-caching \
--attention-backend ROCM_AITER_UNIFIED_ATTN \
--moe-backend aiter \
--kv-cache-dtype fp8 \
--default-chat-template-kwargs '{"thinking_mode":"enabled"}' \
--max-num-seqs 32 \
--speculative-config '{"method": "eagle3", "model": "Inferact/MiniMax-M3-EAGLE3-GQA", "num_speculative_tokens": 3, "attention_backend": "ROCM_AITER_UNIFIED_ATTN", "rejection_sample_method": "synthetic", "synthetic_acceptance_length": 2.78}'
```

### Accuracy
gsm8k before/after (chat completions + thinking mode, greedy, EAGLE3 with real rejection sampling):

Before:
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.9674|±  |0.0049|
|     |       |strict-match    |     5|exact_match|↑  |0.9674|±  |0.0049|

After:
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.9629|±  |0.0052|
|     |       |strict-match    |     5|exact_match|↑  |0.9636|±  |0.0052|

The delta is within one stderr; router logits are fp32-exact to 2e-7 and the top-8 selection is identical per row, and greedy gsm8k on this serving stack has run-to-run wobble of this size from MoE routing / GEMM heuristic nondeterminism independent of this change.

The gsm8k pair was run on the earlier revision of this PR against the `F.linear` path, before #54845 landed; this PR does not change the kernel's arithmetic, only which program computes which rows and the tile sizes it walks K in.

<details>
<summary>Earlier revision of this PR, before #54845 landed</summary>

This PR originally added a separate ROCm `GateLinear` tier, since fp32 router weights fell through to `F.linear`: a bf16->fp32 upcast copy, a hipBLASLt fp32 GEMM and its split-K reduction, three kernels per layer at decode. #54845 covers that, so the tier is gone and only the two changes above remain. The measurements below are that earlier revision against the `F.linear` path, i.e. they size the fast path as a whole rather than this diff.

torch profile at conc 1, before: three kernels per layer
image

after: one Triton kernel
image

Decode profile, in-service kernel medians, rank 0 trace: conc 1 (M=4) 4.4 + 6.9 + 4.6 = 15.9us -> 4.7us, per-layer span 166.1 -> 153.8us; conc 16 (M=64) 4.7 + 7.0 + 5.8 = 17.5us -> 9.0us, per-layer span 265.7 -> 258.7us. Adjacent kernels unchanged.

MiniMax-M3-MXFP4, MI355X x4 TP4, ISL 32768 / OSL 512, EAGLE3 x3 speculative decode at golden-pinned acceptance length 2.78, median metrics:

|  | conc=1 | conc=16 |
|---|---|---|
| TPOT (ms) | 3.31 -> 3.14 (-5.3%) | 7.25 -> 6.97 (-3.9%) |
| Interactivity (tok/s/user) | 301.8 -> 318.7 (+5.6%) | 137.9 -> 143.5 (+4.0%) |
| E2E latency (s) | 2.37 -> 2.28 (-3.7%) | 4.07 -> 4.00 (-1.7%) |

</details>

<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

### CODE REVIEW

**Comment by benenzhu** [comment]
Found duplicate with https://github.com/vllm-project/vllm/pull/54845

**Comment by Fangzhou-Ai** [comment]
Hi @benenzhu can you take a look at my reply at https://github.com/vllm-project/vllm/pull/54845#discussion_r3931183915 first? Also I want to make sure if this PR is not a duplicate of https://github.com/vllm-project/vllm/pull/55235 .

**Comment by benenzhu** [comment]
> Hi @benenzhu can you take a look at my reply at [#54845 (comment)](https://github.com/vllm-project/vllm/pull/54845#discussion_r3931183915) first? Also I want to make sure if this PR is not a duplicate of #55235 .

@Fangzhou-Ai Hi, thanks, other 4 shapes all tested, and can also have some speed up, I supplement it in the `more shapes` paragraph.

This PR mainly let `M=64` which is `conc=16 with MTP=3` can use the custom router trition kernel too, which can have `-6.60us/layer `. It's not a duplicate of the #55235 which is a topk kernel change.
